dataStructures/queues/queue.go

```diff
@@ -0,0 +1,74 @@
+package main
+
+import (
+	"fmt"
+)
+
+// Queue data structure, it will be a list of Orders
+type Queue []*Order
+
+// Order data type
+type Order struct {
+	priority     int
+	quantity     int
+	product      string
+	customerName string
+}
+
+// New method creates a new order with required properties
+func (order *Order) New(priority, quantity int, product, customerName string) {
+	order.priority = priority
+	order.quantity = quantity
+	order.product = product
+	order.customerName = customerName
+}
+
+// Add method takes an order and adds it to the queue based on priority
+func (queue *Queue) Add(order *Order) {
+	if len(*queue) == 0 {
+		*queue = append(*queue, order)
+	} else {
+		appended := false
+
+		for i, existingOrder := range *queue {
```

Comment:
Have you studied big-O complexity in algorithms yet? What is the complexity of your search for the insert position? Can you make it faster? Let me know if you want a hint :)

Reply:
@aschlosberg Yes, I have studied Big-O complexity, though I feel there's more to know for me to master it.

These are my thoughts on this, let me know what you think. Currently, the big-O complexity is O(n), the operation in the loop will be executed as long as the queue `q` is.

These are my suggestions:
- break out of the loop when I get an order that has a lesser priority than the current order, that still makes the complexity O(n)
- The next thing I can think of to make the insertion faster after reviewing one of my books is to implement the priority queue in form of a binary heap.

What do you think?

Reply:
You're correct on both points :) breaking out the loop will make it shorter but, as you say, still linear / O(n). A heap will indeed reduce the algorithm complexity (do you know what to?). 

There is a simpler approach that will achieve the same complexity as a heap but will be much simpler to implement and will allow you to keep everything else as it is (all you have to do is figure out where to insert it). Let me know if you want a hint!